stan/math/rev/fun/trace_quad_form.hpp

```diff
@@ -161,26 +108,25 @@ inline var trace_quad_form(const Mat1& A, const Mat2& B) {
  */
 template <typename Mat1, typename Mat2,
           require_all_matrix_t<Mat1, Mat2>* = nullptr,
-          require_any_var_matrix_t<Mat1, Mat2>* = nullptr,
-	  require_all_not_eigen_st<is_var, Mat1, Mat2>* = nullptr>
+          require_any_var_matrix_t<Mat1, Mat2>* = nullptr>
 inline var trace_quad_form(const Mat1& A, const Mat2& B) {
   check_square("trace_quad_form", "A", A);
   check_multiplicable("trace_quad_form", "A", A, "B", B);
 
-  arena_t<Mat1> arena_A = A;
-  arena_t<Mat2> arena_B = B;
+  auto arena_A = to_arena(to_var_value_if<!is_constant<Mat1>::value>(A));
+  auto arena_B = to_arena(to_var_value_if<!is_constant<Mat2>::value>(B));
 
   var res
       = (value_of(arena_B).transpose() * value_of(arena_A) * value_of(arena_B))
             .trace();
 
   reverse_pass_callback([arena_A, arena_B, res]() mutable {
     if (!is_constant<Mat1>::value)
-      forward_as<promote_scalar_t<var, Mat1>>(arena_A).adj()
+      forward_as<promote_var_matrix_t<Mat1, Mat1, Mat2>>(arena_A).adj()
```

Comment:
I think we should redo `promote_var_matrix_t` in another pull request.

I think it should just be `promote_var_matrix_t<T>` returns the `var_mat` with the same shape as `T`.

I think our current design of `promote_var_matrix_t` would be useful when writing two argument functions that took all combinations of `var<mat>` and `mat<var>`.

It seems now we are writing functions where the signatures all separate:

One function does `(mat<var>, mat<var>)` and returns `mat<var>`, and then another function handles `(var<mat>, mat<var>)`, `(mat<var>, var<mat>)`, `(var<mat>, var<mat>)` and returns a `var<mat>` type.

So it's easy enough from the separation of functions to know what the return types should be, and so `promote_var_matrix_t` can simplify.

Reply:
I think what your seeing here is moreso because of the pattern your using. imo I'd rather just not use this pattern

Reply:
I think I could just do `var_value<Mat1>` since you mentioned it elsewhere.

Reply:
Oh wait that doesn't work cuz might end up with a `var_value<var_value<...>>`.

Anyway, you're right it's a result of the pattern so it's kinda downstream of the rest of the comments.